overlord/snapstate/target.go

```diff
@@ -1090,10 +1090,16 @@ func validateAndInitStoreUpdates(allSnaps map[string]*SnapState, updates map[str
 			sn.RevOpts.CohortKey = snapst.CohortKey
 		}
 
-		var err error
-		sn.RevOpts.Channel, err = resolveChannel(sn.InstanceName, snapst.TrackingChannel, sn.RevOpts.Channel, opts.DeviceCtx)
-		if err != nil {
-			return err
+		// if either the revision is not set (which means this is just a normal
+		// refresh), or the channel is set, then we need to resolve the channel
+		// to something. if the revison is set and the channel is not set, then
+		// we shouldn't do anything with the channel
+		if sn.RevOpts.Revision.Unset() || sn.RevOpts.Channel != "" {
+			var err error
+			sn.RevOpts.Channel, err = resolveChannel(sn.InstanceName, snapst.TrackingChannel, sn.RevOpts.Channel, opts.DeviceCtx)
+			if err != nil {
+				return err
+			}
```

Comment:
I wonder whether for some of these "if" logics we should add method to RevOpts ?

Reply:
I think this is the only place this specific one is used, and I can't think of a good name for it. Maybe `ShouldResolveChannel`?

Alternatively, we are mutating the channel on the RevOpts. Maybe a `ResolveChannel` method that wraps `resolveChannel` and then overwrites it's own channel?

Reply:
yes, the latter does make sense possibly